File: app/MMarket/js/src/data/Broker.js
// Copyright 08-Mar-2020 ºDeme
// GNU General Public License - V3 <http://www.gnu.org/licenses/>

/**
    Broker operations.
**/
export default class Broker {
  /**
      Calculate fees of Broker + Market.
      @param {number} amount
  **/
  static fees (amount) {
    let broker = 9.75;
    if (amount > 25000) {
      broker = amount * 0.001;
    }

    let bolsa = 4.65 + amount * 0.00012;
    if (amount > 140000) {
      bolsa = 13.4;
    } else if (amount > 70000) {
      bolsa = 9.2 + amount * 0.00003;
    } else if (amount > 35000) {
      bolsa = 6.4 + amount * 0.00007;
    }

    return broker + bolsa;
  }

  /**
      Returns net cost of operation.
      @param {number} stocks
      @param {number} price
  **/
  static buy (stocks, price) {
    const amount = stocks * price;
    return amount + Broker.fees(amount);
  }

  /**
      Returns net incomes of operation.
      @param {number} stocks
      @param {number} price
  **/
  static sell (stocks, price) {
    const amount = stocks * price;
    return amount - Broker.fees(amount);
  }

}

File: app/MMarket/c/src/data/broker.c
// Copyright 18-Jan-2022 ºDeme
// GNU General Public License - V3 <http://www.gnu.org/licenses/>

#include "data/broker.h"

double broker_fees(double amount) {
  amount += amount * 0.0017 + 9.53; // penalty for operating.
  double brk = (amount > 50000) ? amount * 0.001 : 9.75;
	double market = amount * 0.00003 + 0.11; // market fee + market execution
  return brk + market;
}

double broker_buy(int stocks, double price) {
  double amount = stocks * price;
	double tobin = amount * 0.002;
  return amount + broker_fees(amount) + tobin;
}

double broker_sell(int stocks, double price) {
  double amount = stocks * price;
  return amount - broker_fees(amount);
}
